### ISSUE TITLE
`include` matcher does not compose with nested matchers for matching keys

### ISSUE THREAD

**Issue by jasonkarns** [rspec/rspec-expectations#1413]
### Subject of the issue

The built-in `include` matcher very nicely supports composing matchers where the values of nested objects within the array/hash can be "recursively" matched against other arbitrary matchers.

However, when matching against `keys`, the `include` matcher directly invokes `.downcase` on the expected values so it doesn't allow composing with other matchers, like `match` for regex matching.

### Your environment
* Ruby version: `ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [arm64-darwin22]`
* rspec-expectations version: `3.12.2`

### Steps to reproduce

```
    expect(last_response.headers)
      .not_to include(
        match(/content/i),
        "Access-Control-Allow-Origin", "Access-Control-Allow-Methods"
      )
```

### Expected behavior

I'd expect the test to fail if a key is found that matches `content` (case-insensitively).

### Actual behavior

NoMethodError

```
  1) cors allows dev-secure origins
     Failure/Error:
       expect(last_response.headers)
         .not_to include(
           match(/content/),
           "Access-Control-Allow-Origin", "Access-Control-Allow-Methods"
         )
     
     NoMethodError:
       undefined method `downcase' for #<RSpec::Matchers::BuiltIn::Match:0x0000000111435f78 @expected=/content/, @expected_captures=nil>
     
               super || @names.include?(k.downcase)
                                         ^^^^^^^^^
```

**Comment by jasonkarns** [comment]
The docs do indicate that matching against keys via regex is supported. (from https://rspec.info/features/3-12/rspec-expectations/built-in-matchers/include/):

```
expect(:a => 1, :b => 2).to include(match(/b/) => 2)
expect(:a => 1, :b => 2).to include(match(/b/) => be_even)
```

And this works, provided one passes a hash to `include` instead of multiple args.

**Comment by vinnydiehl** [comment]
I'm unable to reproduce this. Perhaps I'm misunderstanding. Are these cases representative of your issue?
```rb
RSpec.describe '#include matcher' do
  context 'when composing a #match matcher with multiple arguments' do
    it 'passes when the keys do not contain the matcher' do
      expect(foo: 1, bar: 2, baz: 3).not_to include(
        match(/content/i),
        'Access-Control-Allow-Origin',
        'Access-Control-Allow-Methods'
      )
    end

    it 'fails when the keys contain the matcher' do
      expect {
        expect(Content_test: 1, foo: 2, baz: 3).not_to include(
          match(/content/i),
          'Access-Control-Allow-Origin',
          'Access-Control-Allow-Methods'
        )
      }.to raise_error(RSpec::Expectations::ExpectationNotMetError, /Diff:/)
    end
  end
end
```
An example value of `last_response.headers` that is causing this `NoMethodError` for you would be helpful.

**Comment by JonRowe** [comment]
So your last point made me think some more, the fact that it mostly works suggests there was something that could be improved, it turns out we already have fallback logic for `key?` not having a direct lookup to allow fuzzy matching in the first place (its just slow) and thus #1416 should fix your issue and I think is a fair improvement.

**Comment by jasonkarns** [comment]
Thanks so much, I'm excited to give it a shot!

**Comment by JonRowe** [comment]
(This was released in 3.12.3)

**Comment by jasonkarns** [comment]
🚀 bumped our version and successfully trimmed our specs to assert (case-insensitively!) the desired keys are missing. Thanks again!
